### PR TITLE
Make ActiveAdmin Rails 6 compatible

### LINKED ISSUE #5558: Road to Rails 6
Rails 6 will [use Webpacker as the default](https://github.com/rails/rails/pull/33079). Supporting this can also be an opportunity to modernize ActiveAdmin assets, for distribution by Webpacker. This could also fall in line with the ActiveAdmin 2.0 release.

Some discussion on assets has been [made here](https://github.com/activeadmin/activeadmin/issues/3392)

My suggestion is that we move all assets (CSS and JS) into a 'theme', so that the first theme would be ActiveAdmin Classic, which uses jQuery and the current styling, but which is served via Webpack.

Taking this step would then allow the easy creation of new themes, for example with different styling and completely different JS, dropping jQuery if desired. I suggest the main repo would being in the ActiveAdmin Base and the default theme of ActiveAdmin Classic, which would live in an independent repo.

### PR DESCRIPTION
**Pull request by ionut998**
Opening this PR to get some opinions and ideas about making ActiveAdmin Rails 6 compatible
I used a few unreleased versions of gems that are compatible with Rails 6, but I will change to the official release once that's released.
Most of the tests pass and I am trying now to fix the rest. I tested an existent application against this version of ActiveAdmin and it looks like everything works fine.

Closes #5558.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -22,7 +22,9 @@ group :docs do
   gem 'kramdown' # Markdown implementation (for yard)
 end
 
-gem "rails", "~> 5.2.2"
-gem "activerecord-jdbcsqlite3-adapter", "~> 52.0", platform: :jruby
+# Use the unreleased version of inherited_resources because is compatible with Rails 6
+gem 'inherited_resources', github: 'activeadmin/inherited_resources'
+gem "rails", "6.0.0.beta2"
+gem "activerecord-jdbcsqlite3-adapter", github: 'jruby/activerecord-jdbc-adapter', platform: :jruby
```
Comment (deivid-rodriguez) [inline Gemfile]: I think we can follow the same strategy as in https://github.com/activeadmin/inherited_resources/pull/479. Let's leave the main Gemfile intact for now, and tackle Rails 6.0 on a different gemfile, for example, `gemfiles/rails_60.gemfile{,.lock}`.

Comment (ionut998) [reply Gemfile]: I separated the Gemfile for Rails 6 @deivid-rodriguez

**Comment by yagudaev** [comment]
@ionut998 great work thanks for the PR it is much needed.

I tried using your branch and hit an issue when creating generators with rails 6 as follow:

```
my-app (scaffold-models) $ rails g model User first_name:string last_name:string email:string 
/Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:76:in `block in load_missing_constant': uninitialized constant ActionController::Responder (NameError)
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:8:in `without_bootsnap_cache'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:76:in `rescue in load_missing_constant'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/active_support.rb:58:in `load_missing_constant'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/inherited_resources-1.3.1/lib/inherited_resources/responder.rb:2:in `<module:InheritedResources>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/inherited_resources-1.3.1/lib/inherited_resources/responder.rb:1:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/inherited_resources-1.3.1/lib/inherited_resources.rb:4:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/bundler/gems/activeadmin-d935f2b13efc/lib/active_admin.rb:9:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/bundler/gems/activeadmin-d935f2b13efc/lib/activeadmin.rb:1:in `<main>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `block in require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:83:in `register'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:20:in `require_with_bootsnap_lfi'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:29:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:81:in `block (2 levels) in require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:76:in `each'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:76:in `block in require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:65:in `each'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/runtime.rb:65:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler.rb:114:in `require'
        from /Users/mike/projects/my/my-app/config/application.rb:20:in `<top (required)>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:92:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:92:in `preload'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:153:in `serve'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:141:in `block in run'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:135:in `loop'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application.rb:135:in `run'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/gems/2.6.0/gems/spring-2.0.2/lib/spring/application/boot.rb:19:in `<top (required)>'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
        from /Users/mike/.rbenv/versions/2.6.1/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
        from -e:1:in `<main>'
```

File: activeadmin.gemspec [hunk 2]
```diff
@@ -22,10 +22,10 @@ Gem::Specification.new do |s|
   s.add_dependency 'arbre', '~> 1.1', '>= 1.1.1'
   s.add_dependency 'formtastic', '~> 3.1'
   s.add_dependency 'formtastic_i18n', '~> 0.4'
-  s.add_dependency 'inherited_resources', '~> 1.7'
+  s.add_dependency 'inherited_resources'
```
Comment (deivid-rodriguez) [inline activeadmin.gemspec]: Is this change needed? Master branch of inherited resources is still compatible with this requirement I think.

File: activeadmin.gemspec [hunk 3]
```diff
@@ -22,10 +22,10 @@ Gem::Specification.new do |s|
...
   s.add_dependency 'jquery-rails', '~> 4.2'
   s.add_dependency 'kaminari', '~> 1.0', '>= 1.0.1'
-  s.add_dependency 'railties', '>= 5.0', '< 5.3'
+  s.add_dependency 'railties', '>= 5.0'
```
Comment (deivid-rodriguez) [inline activeadmin.gemspec]: Let's keep an upper limit here. I think `< 6.0` to allow 6.0 prereleases works for now.

**Comment by deivid-rodriguez** [comment]
@yagudaev You need to manually add `gem "inherited_resources", git: "https://github.com/activeadmin/inherited_resources"` to your `Gemfile` for now. Otherwise `bundler` will resolve `inherited_resources` to a very old version (which had no requirement on the rails version) and you'll get that weird error.

File: activeadmin.gemspec (hunk 2 above)
Comment (ionut998) [reply activeadmin.gemspec]: yes, this is actually not needed
I added it back

File: activeadmin.gemspec (hunk 3 above)
Comment (ionut998) [reply activeadmin.gemspec]: I changed to keep the upper limit as `< 6.0`

File: gemfiles/rails_60.gemfile [hunk 4]
```diff
@@ -0,0 +1,12 @@
+source "https://rubygems.org"
+
+eval_gemfile(File.expand_path(File.join("..", "Gemfile.common"), __dir__))
+
+gem "rails", "6.0.0.beta2"
+
+# Use the unreleased version of gems because they are compatible with Rails 6
+gem 'inherited_resources', github: 'activeadmin/inherited_resources'
```
Comment (deivid-rodriguez) [inline gemfiles/rails_60.gemfile]:
```suggestion
gem 'inherited_resources', git: 'https://github.com/activeadmin/inherited_resources'
```

To skip bundler warnings.

File: gemfiles/rails_60.gemfile [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+gem "activerecord-jdbcsqlite3-adapter", github: 'jruby/activerecord-jdbc-adapter', platform: :jruby
```
Comment (deivid-rodriguez) [inline gemfiles/rails_60.gemfile]:
```suggestion
gem "activerecord-jdbcsqlite3-adapter", git: 'https://github.com/jruby/activerecord-jdbc-adapter', platform: :jruby
```

To skip more bundler warnings.

File: gemfiles/rails_60.gemfile [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+gem 'sqlite3', '~> 1.3.6', platform: :mri, group: :test
```
Comment (deivid-rodriguez) [inline gemfiles/rails_60.gemfile]: Remove this line since you have already locked the dependency in `Gemfile.common`.

**Comment by deivid-rodriguez** [comment]
By the way, I don't know why CircleCI logs are "swallowing" cucumber errors. Something to do with the `parallel_tests` setup. If you run the failing features locally you'll see the actual errors.

**Comment by ionut998** [comment]
That last test seem like a tricky one

The actual error is `Invalid single-table inheritance type: Publisher is not a subclass of User (ActiveRecord::SubclassNotFound)` from [this test](https://github.com/activeadmin/activeadmin/blob/master/features/comments/commenting.feature#L134)

Interestingly if I only run this test then it passes. Something is being set by the [test before](https://github.com/activeadmin/activeadmin/blob/master/features/comments/commenting.feature#L118) that make it fail. If I delete the `Commenting on a STI superclass` test then the `Commenting on a STI subclass` test passes.

If I follow the trace and look into Rails it leads me to [this comparison](https://github.com/rails/rails/blob/6a1a1e66ea7a917942bd8369fa8dbfedce391dab/activerecord/lib/active_record/inheritance.rb#L242) that returns false when it shouldn't 
These are the values when this comparison happens:
```
self = User(id: integer, type: string...
subclass = Publisher(id: integer, type: string...
descendants=[User::VIP(id: integer, type: string...), Publisher(id: integer, type: string...)]
```
and strangely when I do `descendants.include?(subclass)` I get false

Any ideas @deivid-rodriguez

**Comment by deivid-rodriguez** [comment]
@ionut998 Seems like a code (re)loading issue... :thinking:. Rails 6 will introduce [a new code loader](https://github.com/fxn/zeitwerk) (which looks awesome, by the way). Maybe you can try setting `config.autoloader = :classic` in the test application. If it passes, we know it's related to the new loader.

**Comment by ionut998** [comment]
Thanks @deivid-rodriguez 
Setting the autoloader to `classic` fixes the issue
I followed the documentation from [here](https://weblog.rubyonrails.org/2019/2/22/zeitwerk-integration-in-rails-6-beta-2/#stis) and added the extra initializer to preload the leafs of the STI but I still get that error.

**Comment by deivid-rodriguez** [comment]
To get the `upload_coverage` job green, you need to adjust the total never of jobs here (sum one to the 19 per job being added in this PR):

https://github.com/activeadmin/activeadmin/blob/17988a9ecd85b144529501dbf2a0c0e9ac09c53b/.circleci/config.yml#L65

File: .circleci/config.yml [hunk 7]
```diff
@@ -208,6 +208,18 @@ jobs:
...
+    docker:
+    - image: circleci/ruby:2.6.1-node
+      user: root
+
+    environment:
+      BUNDLE_GEMFILE: gemfiles/rails_60.gemfile
+
+    steps:
+      *test_app_steps
+
```
Comment (deivid-rodriguez) [inline .circleci/config.yml]: Double blank line :smiley:

File: spec/support/rails_template.rb [hunk 8]
```diff
@@ -114,6 +114,8 @@ class Tagging < ActiveRecord::Base
 
 RUBY
 
+inject_into_file 'config/environments/test.rb', "\n  config.autoloader = :classic\n", after: 'Rails.application.configure do' if Rails::VERSION::MAJOR >= 6
```
Comment (deivid-rodriguez) [inline spec/support/rails_template.rb]: Can we add a TODO note about fixing this in the future, maybe with a reference to the STI cucumber feature that fails with `zeitwerk`?

File: .circleci/config.yml (hunk 7 above)
Comment (ionut998) [reply .circleci/config.yml]: I have removed the extra line

File: spec/support/rails_template.rb (hunk 8 above)
Comment (ionut998) [reply spec/support/rails_template.rb]: I added the TODO note

**Comment by deivid-rodriguez** [comment]
@ionut998 Thanks for doing this! If you can update to the recently released 6.0.0.beta3 since we are at it, other than that it looks great to me!

**Comment by ionut998** [comment]
Changed to Rails 6 beta3 @deivid-rodriguez

File: .circleci/config.yml [hunk 9]
```diff
@@ -62,7 +62,7 @@ version: 2.1
   run:
     name: Upload coverage results to Code Climate
     command: |
-      coverage/test-reporter sum-coverage coverage/codeclimate.*.json --parts 19 --output coverage/codeclimate.total.json
+      coverage/test-reporter sum-coverage coverage/codeclimate.*.json --parts 22 --output coverage/codeclimate.total.json
```
Comment (javierjulio) [inline .circleci/config.yml]: @ionut998 would you mind adding a comment above this? @deivid-rodriguez I saw your comment about this change and didn't remember what it was, thanks. Is the parts number the total count of builds/jobs plus 1? Is the plus 1 the lint job? Maybe a note like:

```
# The parts number is the total jobs count plus 1 (for the lint job)
coverage/test-reporter sum-coverage coverage/codeclimate.*.json --parts 22 --output coverage/codeclimate.total.json
```

Comment (deivid-rodriguez) [reply .circleci/config.yml]: Yeah, that's it. It's the number of jobs that generate coverage reports.

**Comment by ionut998** [comment]
Happy to help @javierjulio

File: .circleci/config.yml (hunk 9 above)
Comment (ionut998) [reply .circleci/config.yml]: Added the comment

**Comment by deivid-rodriguez** [comment]
@ionut998 Sorry for the many requests. I noticed that with the upgrade to beta3, also `zeitwerk` got upgraded to `1.3.3`. Who knows whether this bug fix release could fix our problems with the new loader... Mind giving it a try?

**Comment by ionut998** [comment]
No problem @deivid-rodriguez 
I actually tried before pushing the upgrade. I removed the `autoloader = :classic` from the rails6_beta3 test application and added the custom initializer for preloading the STI leafs but it still errored in the same way as before.

**Comment by javierjulio** [comment]
@ionut998 thanks again for doing this! We all really appreciate it. ❤️

**Comment by deivid-rodriguez** [comment]
Funny that the commit number 6000 in the repo has added support for Rails 6.0 :smiley:.

**Comment by javierjulio** [comment]
Wow, you're right! So cool. Let the record show!

activeadmin-6000-commits-for-rails6-support

**Comment by ionut998** [comment]
Perfect timing for Rails 6 support 😃
